### PR TITLE
Update to use handles in the embedding api

### PR DESCRIPTION
**Pull request by bholmes**
Change ToManagedRepresentation and ToNativeRepresentation to use GHHandles

### CODE REVIEW

File: unity/unity-embed-host/UnsafeExtensions.cs [hunk 1]
```diff
@@ -16,43 +16,28 @@ namespace Unity.CoreCLRHelpers;
...
+        // legacy pointer code path until we are fully GC safe
+        return Unsafe.As<nint, object>(ref intPtr);
     }
 
-#if TESTING_UNITY_CORECLR
-    static void WorkaroundToGetGCSafety(object obj)
+    public static nint ToNativeRepresentation(this object obj)
     {
-        var valueToPin = 1;
-        var handle = GCHandle.Alloc(valueToPin, GCHandleType.Pinned);
```
Comment (joncham) [inline unity/unity-embed-host/UnsafeExtensions.cs:38]: I see these were previously pinned, but I don't recall why that was needed @mrvoorhe ?

**Comment by joncham** [comment]
After talking to the performance team, I think we'd like to enable GC safety for the CoreCLR Player but not the Editor. 

First, I think we need to make the real GC a runtime check.  Likely via the call to `mono_jit_init_version` we could use the passed in version string to decide if we want to enable the real GC or load our existing Null GC.

As far as what representation is passed across the embedding API boundary. I think we have two options:

1. Enable GC safety handles here always. Return a GC handle for object return values. Accept either a pointer or a GC handle. Then on the Unity side at the embedding layer, resolve the GCHandle return values when in Editor.
2. Make the mode here a runtime switch. Only return handles if the GC is enable via the runtime switch above.

Thoughts?

**Comment by joncham** [comment]
Restating the above, I think the goal should be to be able to land the next runtime binary zip in Unity with the ability to turn the real GC on/off at runtime.

File: unity/unity-embed-host/UnsafeExtensions.cs (hunk 1 above)
Comment (mrvoorhe) [reply unity/unity-embed-host/UnsafeExtensions.cs:38]: Because we cached these in a static concurrent bag to ensure GC safety during tests.

**Review by mrvoorhe — APPROVED** [review]

**Comment by mrvoorhe** [comment]
> After talking to the performance team, I think we'd like to enable GC safety for the CoreCLR Player but not the Editor.
> 
> First, I think we need to make the real GC a runtime check. Likely via the call to `mono_jit_init_version` we could use the passed in version string to decide if we want to enable the real GC or load our existing Null GC.
> 
> As far as what representation is passed across the embedding API boundary. I think we have two options:
> 
> 1. Enable GC safety handles here always. Return a GC handle for object return values. Accept either a pointer or a GC handle. Then on the Unity side at the embedding layer, resolve the GCHandle return values when in Editor.
> 2. Make the mode here a runtime switch. Only return handles if the GC is enable via the runtime switch above.
> 
> Thoughts?

It seems like (2) is less work?

File: unity/unity-embed-host/UnsafeExtensions.cs (hunk 1 above)
Comment (bholmes) [reply unity/unity-embed-host/UnsafeExtensions.cs:38]: I think we are fine to remove this. With the UnityGC I this should not be needed.  When CoreCLR GC is enabled, we need to fix the tests anyway.  + The tests may be replaced with managed soon.

File: unity/unity-embed-host/UnsafeExtensions.cs [hunk 2]
```diff
@@ -16,43 +16,34 @@ namespace Unity.CoreCLRHelpers;
 /// </summary>
 static class UnsafeExtensions
 {
-#if TESTING_UNITY_CORECLR
-    private static readonly ConcurrentBag<GCHandle> handles = new ();
-#endif
-
+    private static readonly bool s_ReturnHandlesFromAPI = CoreCLRHost.UseRealGC() || CoreCLRHost.ReturnHandlesFromAPI();
```
Comment (mrvoorhe) [inline unity/unity-embed-host/UnsafeExtensions.cs:19]: Nit: We can use normal C# naming convention here. `_returnHandlesFromAPI`

File: src/coreclr/vm/mono/mono_coreclr.cpp [hunk 3]
```diff
@@ -1688,15 +1706,26 @@ extern "C" EXPORT_API void EXPORT_CC mono_unity_initialize_host_apis(initialize_
 
 extern "C" EXPORT_API MonoDomain* EXPORT_CC mono_jit_init_version(const char *file, const char* runtime_version)
 {
+    if (runtime_version != NULL)
+    {
+        if (strstr(runtime_version, "return-handles-from-api"))
+            s_ReturnHandlesFromAPI = true;
```
Comment (mrvoorhe) [inline src/coreclr/vm/mono/mono_coreclr.cpp:1712]: I think enabling by default and then having a switch to turn off would be better.  A few reasons
(1) Correct me if I'm wrong, but I think with this opt-in change the tests are not running with the real GC path.  Keeping the embedding api tests green with the real gc seems like a minimum bar before even bothering with the real gc in unity.
(2) Opt-in means that we'll have to keep `use-real-gc` in Unity until we come back here and remove this code.  If we turn it on and go with opt-out, then we can add `disable-real-gc` for this interim period while we get it turned on and once it's ready remove `disable-real-gc`

File: unity/unity-embed-host/UnsafeExtensions.cs (hunk 1 above)
Comment (mrvoorhe) [reply unity/unity-embed-host/UnsafeExtensions.cs:38]: I missed your comment @bholmes .  This is for the managed embedding tests that call into native.  I wasn't able to get the null gc working for linux and mac.  https://github.com/Unity-Technologies/runtime/blob/81ca339de14bd78c1f527946f5161c604554c491/unity/CITools/BuildDriver/EmbeddingHost.cs#L201

**Pull request merged**